src/hotspot/share/opto/phaseX.cpp

```diff
@@ -2269,6 +2269,41 @@ void PhasePeephole::print_statistics() {
 }
 #endif
 
+Node* PhaseLowering::apply_ideal(Node* k, bool can_reshape) {
+  // Run the lowered Ideal method to continue doing transformations on the node, while avoiding existing transforms
+  // that may undo the changes done during lowering.
+  return k->LoweredIdeal(this);
+}
+
+Node* PhaseLowering::lower_node(Node* n) {
+  // Apply shared lowering transforms
+
+  // Apply backend-specific lowering transforms
+  return lower_node_platform(n);
+}
+
+void PhaseLowering::lower() {
+  // Worklist should be empty before lowering
+  _worklist.ensure_empty();
+
+  C->identify_useful_nodes(_worklist);
```

Comment:
To address @iwanowww 's concern, you can have a backend-specific method `PhaseLowering::do_lowering()` that will decide whether we should perform lowering on a particular graph.

Reply:
I think if we're planning on moving optimizations like `DivMod` to the platform-independent part of this phase we may end up processing nodes more generally, since the graph patterns we expect would be pretty wide and shared between multiple backends. What do you think?

Reply:
But there are still backends where we don't do such optimization, so it is still worth it to have a predicate for it.